parsers/BD.py

```diff
@@ -1,299 +1,295 @@
-    for index, row in df.iterrows():
-        production = row.to_dict()
-        dt = timestamp_converter(index, target_datetime)
-
-        mapped_production = {
-            MAPPING[k]: v for k, v in production.items() if k in MAPPING
-        }
-        mapped_production["gas"] = mapped_production.pop(
-            "gas_public"
-        ) + mapped_production.pop("gas_private")
-        mapped_production["oil"] = mapped_production.pop(
-            "oil_public"
-        ) + mapped_production.pop("oil_private")
-        processed_data.append((mapped_production, dt))
+    rows = table_body.find_all("tr")
+    row_data = []
+
+    for row in rows:
+        row_items = row.find_all("td")
+        row_items = [item.text.strip() for item in row_items]
+
+        row_dict = dict()
+
+        # date and time in [0]; [1] are DD-MM-YYYY; hh:mm:ss
+        row_dict["time"] = arrow.get(
+            row_items[0] + row_items[1], "DD-MM-YYYY" + "HH:mm:ss", tzinfo=tz
+        ).datetime
+        # [2] is total generation in MW
+        row_dict["total_generation"] = table_entry_to_float(row_items[2])
+        # [3] is total demand in MW
+        row_dict["total_demand"] = table_entry_to_float(row_items[3])
+        # [4] is loadshed, not interesting for us
+        row_dict["loadshed"] = table_entry_to_float(row_items[4])
+        # [5] is generation from gas
+        row_dict["gas"] = table_entry_to_float(row_items[5])
+        # [6] is generation from liquid fuel (oil)
+        row_dict["oil"] = table_entry_to_float(row_items[6])
+        # [7] is generation from coal
+        row_dict["coal"] = table_entry_to_float(row_items[7])
+        # [8] is generation from hydro
+        row_dict["hydro"] = table_entry_to_float(row_items[8])
+        # [9] is generation from solar
+        row_dict["solar"] = table_entry_to_float(row_items[9])
+        # [10], [11] is import from Bheramara PP (IN-EA) and Tripura (IN-NE)
+        row_dict["bd_import_bheramara"] = table_entry_to_float(row_items[10])
+        row_dict["bd_import_tripura"] = table_entry_to_float(row_items[11])
+        row_dict["remarks"] = row_items[12]
+
+        row_data.append(row_dict)
+
+    return row_data
+
+
+def verify_table(table_header: Tag):
+    """
+    Validate the table headers, that it looks like expected.
+    Don't parse if the table has changed.
+    """
+    header_items = table_header.find_all("th")
+    header_items = [item.text.strip() for item in header_items]
 
-    return processed_data
+    if header_items != TABLE_HEADERS:
+        raise ParserException(
+            parser="BD.py",
+            message=(
+                f"Table headers mismatch with expected ones."
+                f"Expected: {TABLE_HEADERS}"
+                f"Parsed: {header_items}"
+            ),
+        )
 
 
-def exchange_processer(df, target_datetime, old_format=False) -> list:
+def query(
+    session: Optional[Session], target_datetime: Optional[datetime], logger: Logger
```

Comment:
```suggestion
    session: Session, target_datetime: Optional[datetime], logger: Logger
```
Session shoud be defined here to avoid type errors further down in the code.